pkg/agent/ipassigner/responder/ndp_responder.go

```diff
@@ -108,27 +101,111 @@ func (r *ndpResponder) handleNeighborSolicitation() error {
 		Options: []ndp.Option{
 			&ndp.LinkLayerAddress{
 				Direction: ndp.Target,
-				Addr:      r.iface.HardwareAddr,
+				Addr:      link.HardwareAddr,
 			},
 		},
 	}
-	if err := r.conn.WriteTo(na, nil, srcIP); err != nil {
+	if err := conn.WriteTo(na, nil, srcIP); err != nil {
 		return err
 	}
-	klog.V(4).InfoS("Sent Neighbor Advertisement", "ip", ns.TargetAddress.String(), "interface", r.iface.Name)
+	klog.V(4).InfoS("Sent Neighbor Advertisement", "ip", ns.TargetAddress.String(), "interface", r.linkName)
 	return nil
 }
 
 func (r *ndpResponder) Run(stopCh <-chan struct{}) {
+	// The responder instance is created by the factory and can be shared by multiple callers.
+	// Using once.Do here ensures it is started only once.
+	r.once.Do(func() {
+		wait.NonSlidingUntil(func() {
+			r.dialAndHandleRequests(stopCh)
+		}, time.Second, stopCh)
+	})
+	<-stopCh
+}
+
+func (r *ndpResponder) dialAndHandleRequests(endCh <-chan struct{}) {
+	transportInterface, err := net.InterfaceByName(r.linkName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get interface", "interface", r.linkName)
+		return
+	}
+	addrs, err := transportInterface.Addrs()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get addresses for interface", "interface", r.linkName)
+		return
+	}
+	var ip netip.Addr
+	for _, addr := range addrs {
+		if ipnet, ok := addr.(*net.IPNet); ok {
+			if utilnet.IsIPv6(ipnet.IP) && ipnet.IP.IsLinkLocalUnicast() {
+				ip, _ = netip.AddrFromSlice(ipnet.IP)
+				break
+			}
+		}
+	}
+	if ip.IsUnspecified() {
+		klog.InfoS("Interface does not have a link-local IPv6 address yet", "interface", r.linkName)
```

Comment:
could it overwhelm log file if the interface doesn't have link-local address? Is it rare to not have link-local IPv6 address?

Reply:
the check here is not needed anymore. I have removed the changes.